hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseHoodieWriteClient.java

```diff
@@ -940,46 +954,53 @@ public String startCommit(String actionType, HoodieTableMetaClient metaClient) {
     if (needsUpgrade(metaClient)) {
       // unclear what instant to use, since upgrade does have a given instant.
       executeUsingTxnManager(Option.empty(), () -> tryUpgrade(metaClient, Option.empty()));
     }
     CleanerUtils.rollbackFailedWrites(config.getFailedWritesCleanPolicy(),
         HoodieTimeline.COMMIT_ACTION, () -> tableServiceClient.rollbackFailedWrites(metaClient));
 
-    LOG.info("Generate a new instant time: {} action: {}", instantTime, actionType);
-    // check there are no inflight restore before starting a new commit.
-    HoodieTimeline inflightRestoreTimeline = metaClient.getActiveTimeline().getRestoreTimeline().filterInflightsAndRequested();
-    ValidationUtils.checkArgument(inflightRestoreTimeline.countInstants() == 0,
-        () -> "Found pending restore in active timeline. Please complete the restore fully before proceeding. As of now, "
-            + "table could be in an inconsistent state. Pending restores: "
-            + Arrays.toString(inflightRestoreTimeline.getInstantsAsStream().map(HoodieInstant::requestedTime).toArray()));
-
-    if (config.getFailedWritesCleanPolicy().isLazy()) {
-      this.heartbeatClient.start(instantTime);
-    }
+    txnManager.beginTransaction(Option.empty(), Option.empty());
+    String instantTime;
+    try {
+      instantTime = providedInstantTime.orElseGet(() -> createNewInstantTime(false));
+      LOG.info("Generate a new instant time: {} action: {}", instantTime, actionType);
+      // check there are no inflight restore before starting a new commit.
+      HoodieTimeline inflightRestoreTimeline = metaClient.reloadActiveTimeline().getRestoreTimeline().filterInflightsAndRequested();
+      ValidationUtils.checkArgument(inflightRestoreTimeline.countInstants() == 0,
+          () -> "Found pending restore in active timeline. Please complete the restore fully before proceeding. As of now, "
+              + "table could be in an inconsistent state. Pending restores: "
+              + Arrays.toString(inflightRestoreTimeline.getInstantsAsStream().map(HoodieInstant::requestedTime).toArray()));
+
+      if (config.getFailedWritesCleanPolicy().isLazy()) {
+        this.heartbeatClient.start(instantTime);
+      }
 
-    if (ClusteringUtils.isClusteringOrReplaceCommitAction(actionType)) {
-      metaClient.getActiveTimeline().createRequestedCommitWithReplaceMetadata(instantTime, actionType);
-    } else {
-      metaClient.getActiveTimeline().createNewInstant(metaClient.createNewInstant(HoodieInstant.State.REQUESTED, actionType,
-          instantTime));
+      if (ClusteringUtils.isClusteringOrReplaceCommitAction(actionType)) {
+        metaClient.getActiveTimeline().createRequestedCommitWithReplaceMetadata(instantTime, actionType);
+      } else {
+        metaClient.getActiveTimeline().createNewInstant(metaClient.createNewInstant(HoodieInstant.State.REQUESTED, actionType, instantTime));
```

Comment:
Previous assumption is that it is OK for `t1.request` to be created on storage after `t2.request` is created on storage, so there is no need for locking here.

Reply:
yeah, the fix is not necessary, the requested instant just behaves like a marker to index files/completion time, there is no strong requirement that the real creation time must follows the same sequence.

Reply:
The requested time is used in the conflict detection code so this can make it challenging to debug errors that happening due to conflicts or due to failures in conflict detection. If you think about the sequencing of generating the timestamp and then writing it across two writers, you end up with 6 possibilities in the current system because the generation and persistence are separate events vs 2 possibilities if we ensure the timeline is generated and then written in the same block.

Reply:
That design doc does not really reach consensus yet, let's not make eager changes based on it, actually I kind of think within that design, the conflict resolution are scattered to all over the places which is impossible to maintain for a developer. And the ramifications are way too many and also very complicated.

Reply:
Added `DRAFT` to the PR but I am unable to change this back to draft in the GitHub UI. Feel free to also mark as changes requested so we make sure consensus is reached first.

Reply:
> there is no strong requirement that the real creation time must follows the same sequence.

@danny0405 @yihua Can you describe how this is not necessary for NBCC, and compaction scheduling . Don't we want the time generation + requesting to be done to order the log creation and compaction scheduling. i.e requested_time of compaction > completion_time of log - is how we do the slicing no? 

note that @the-other-tim-brown is thinking about deployment scenarios outside of current model, where the write does the compaction scheduling. 

Whats your actual concern with doing this? performance? I'd like to simplify this first. and solve for performance on top later.

Reply:
> NBCC, and compaction scheduling . Don't we want the time generation + requesting to be done to order the log creation and compaction scheduling. i.e requested_time of compaction > completion_time of log

yeah, but the requested instant time of compaction just behaviors like a barrier marker in the timeline history of completion time from log commits, the creation sequence does not affect the correctness.

> Whats your actual concern with doing this? performance?

More locks would induce more blocking on requests, it's just no necessary to do it now on OSS.

Reply:
Currently this is not introducing any more locking steps. There is a lock to create the instant time currently. I am just updating this so we lock to create the time and persist it to avoid race conditions in multi-writer scenarios. For single-writers, they will not have any impact. 

Additionally, for multi-writer cases we will currently require taking the lock twice per table service, once to generate the requested time and once to perform the planning. We can change this to similarly only require a single request to take the lock.